plans/external.pp

```diff
@@ -0,0 +1,57 @@
+# @summary
+#   A plan that generates external facts based on the provided modulepath and
+#   sets facts on specified targets.
+#
+# @param path The path to the directory on localhost containing external facts
+# @param targets The targest the collect and set facts on
+# @return The target objects, with facts added
+# @example Gather external facts from an installed module
+#   $moduledir = module_directory('mymod')
+#   $with_facts = run_plan(facts::external, $targets, path => file::join($moduledir, 'facts.d'))
+#   return $with_facts.map |$target| { $target.facts }
+plan facts::external(
+  String $path,
+  TargetSpec $targets
+) {
+  $t = get_targets($targets)
+
+  # Strip out dotfiles
+  $facts_paths = dir::children($path).filter |$path| { $path[0] != '.' }
+  $facts_paths.each |$file| {
+    $ext_facts = run_script(file::join($path, $file), $t)
+    $t.each |$target| {
+      $raw_facts = $ext_facts.find($target.name).value['stdout']
+      $jsonoryaml = catch_errors() || {
+        parseyaml($raw_facts)
+      }
+
+      # If parsing as YAML or JSON failed, the error will be a string and we
+      # should try to parse the original facts as key-value pairs
+      if type($jsonoryaml, 'generalized') == String {
+        # Split on newlines and equals, then convert to a hash
+        $fact_kv_strings = $raw_facts.split(/\n/)
+        $fact_kvs = $fact_kv_strings.map |$str| {
+          unless $str =~ /\=/ {
+            $msg = @("MSG"/L)
+              External fact output must have key-value pairs separated by an '=' on each line, like so:
+              key=value
+              key2=value2
+              | MSG
+            fail_plan($msg)
+          }
+          $splits = $str.split(/\=/)
+          # Account for facts whose values have an '='
+          $value = $splits[1, -1].join("")
+          [$splits[1], $value]
```

Comment:
This strips any `=` from the value entirely. It should be `join("=")` to put them back. It also returns the first component of the value as the key. So for instance, `key=foo=bar` returns `["foo", "foobar"]` instead of `["key", "foo=bar"]`.

I think this would be simpler as just a regex match. Since you're actually already doing a regex match above for validation, it could probably be folded into that.

```puppet
$match = $str.match(/^([^=]+)=(.+)$/)
if $match {
  $key = $match[1]
  $value = $match[2]
} else {
  fail_plan([...])
}
```

Reply:
Clever, and much simpler! I really just wish `split()` took a limit, but...changing that seems like a lot more work than it's worth.